src/lsm/zig_zag_merge.zig

```diff
@@ -0,0 +1,664 @@
+
+                    if (constants.verify) {
+                        // It's assumed that streams will produce the same value.
+                        assert(stdx.equal_bytes(Value, &value, &value_other));
+                    }
+                }
+
+                if (it.previous_key_popped) |previous| {
+                    switch (std.math.order(previous, key)) {
+                        .lt => assert(it.direction == .ascending),
+                        // Discard duplicate values.
+                        .eq => continue,
+                        .gt => assert(it.direction == .descending),
+                    }
+                }
+                it.previous_key_popped = key;
+
+                return value;
+            }
+
+            return null;
+        }
+
+        fn peek_key(it: *ZigZagMergeIterator) error{Drained}!?Key {
+            assert(it.streams_count <= streams_max);
+            assert(it.streams_count > 1);
+
+            const key_min: Key = switch (it.direction) {
+                .ascending => 0,
+                .descending => std.math.maxInt(Key),
+            };
+
+            var drained: BitSet = BitSet.initEmpty();
+            var probe_key: Key = key_min;
+
+            var probing: BitSet = BitSet.initFull();
+            while (probing.count() > 0) {
+                // Looking into all non-drained streams for a match, while accumulating
+                // the most ahead key to probe the streams behind.
+                probing = BitSet.initEmpty();
+                for (0..it.streams_count) |stream_index| {
+                    if (drained.isSet(stream_index)) continue;
+
+                    const key = stream_peek(it.context, @intCast(stream_index)) catch |err| {
+                        switch (err) {
+                            // Return immediately on empty streams.
+                            error.Empty => return null,
+                            // Skipping `Drained` streams. The goal is to match all buffered streams
+                            // first so that the drained ones can read from a narrower key range.
+                            error.Drained => {
+                                drained.set(stream_index);
+                                continue;
+                            },
+                        }
+                    };
+
+                    if (switch (it.direction) {
+                        .ascending => key > probe_key,
+                        .descending => key < probe_key,
+                    }) {
+                        // The stream is ahead, it will be the probe key,
+                        // meaning all streams before must be probed.
+                        probe_key = key;
+
+                        // Setting all previous streams as `true` except the drained ones.
+                        probing.setRangeValue(.{ .start = 0, .end = stream_index }, true);
+                        probing.setIntersection(drained.complement());
```

Comment:
>Though, this might be more clearly represented as a probe_stream usize/index rather than a probing bitset.

We may have more than one stream as the "seed".

Example:
```
A.peek() == 2
B.peek() == 2
C.peek() == 1
```

The probing bitset will contain just `C`, with `probe_key=2`.